File: src/main/java/net/vpg/vjson/value/JSONArray.java
/*
 * Copyright 2021 Vaibhav Nargwani
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.vpg.vjson.value;

import net.vpg.vjson.SerializableArray;
import net.vpg.vjson.parser.ParseException;
import net.vpg.vjson.pretty.PrettyPrintConfig;
import net.vpg.vjson.pretty.PrettyPrinter;
import net.vpg.vjson.reader.JSONReader;

import java.io.*;
import java.net.URL;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collector;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class JSONArray extends JSONValue implements SerializableArray, JSONContainer<Integer> {
    private final List<JSONValue> list;

    public JSONArray() {
        this.list = new ArrayList<>();
    }

    private JSONArray(List<?> list) {
        this.list = list.stream().map(JSONValue::of).collect(Collectors.toList());
    }

    public static JSONArray of(List<?> list) {
        return new JSONArray(list);
    }

    public static JSONArray parse(Reader in) throws ParseException {
        return getParser().parse(in).toArray();
    }

    public static JSONArray parse(URL url) throws ParseException, IOException {
        return getParser().parse(url).toArray();
    }

    public static JSONArray parse(InputStream in) throws ParseException {
        return getParser().parse(in).toArray();
    }

    public static JSONArray parse(String s) throws ParseException {
        return getParser().parse(s).toArray();
    }

    public static JSONArray parse(File f) throws ParseException, FileNotFoundException {
        return getParser().parse(f).toArray();
    }

    public static JSONArray parse(JSONReader s) throws ParseException {
        return getParser().parse(s).toArray();
    }

    public static JSONArray parseFile(String path) throws ParseException, FileNotFoundException {
        return parse(new File(path));
    }

    public static <T> Collector<T, ?, JSONArray> collector() {
        return Collector.of(JSONArray::new, JSONArray::add, JSONArray::addAll);
    }

    public int size() {
        return list.size();
    }

    public boolean isEmpty() {
        return list.isEmpty();
    }

    public JSONValue get(Integer index) {
        return JSONValue.of(list.get(index));
    }

    public JSONArray add(int index, Object value) {
        list.add(index, JSONValue.of(value));
        return this;
    }

    public JSONArray add(Object value) {
        list.add(JSONValue.of(value));
        return this;
    }

    public JSONArray addAll(Collection<?> values) {
        values.forEach(this::add);
        return this;
    }

    public JSONArray addAll(JSONArray array) {
        list.addAll(array.list);
        return this;
    }

    public JSONArray remove(int index) {
        list.remove(index);
        return this;
    }

    public List<JSONValue> toList() {
        return list;
    }

    public <T> List<T> toList(Function<JSONValue, T> converter) {
        return list.stream()
            .map(converter)
            .collect(Collectors.toList());
    }

    public Stream<JSONValue> stream() {
        return list.stream();
    }

    @Override
    public String deserialize() {
        return list.stream().map(JSONValue::deserialize).collect(Collectors.joining(",", "[", "]"));
    }

    @Override
    public Type getType() {
        return Type.ARRAY;
    }

    @Override
    public Object getRaw() {
        return list.stream().map(JSONValue::getRaw).collect(Collectors.toList());
    }

    @Override
    public JSONArray toArray() {
        return this;
    }

    @Override
    public void toPrettyString(PrettyPrinter printer) {
        PrettyPrintConfig config = printer.getConfig();
        printer.print("[");
        if (config.isArrayContentsOnSameLine()) {
            if (config.isSpaceWithinBrackets())
                printer.space();
        } else {
            printer.incrementIndentLevel();
            printer.newLineAndIndent();
        }
        Iterator<JSONValue> iterator = list.iterator();
        while (iterator.hasNext()) {
            iterator.next().toPrettyString(printer);
            if (iterator.hasNext()) {
                if (config.isSpaceBeforeComma())
                    printer.space();
                printer.print(",");
                if (config.isArrayContentsOnSameLine()) {
                    if (config.isSpaceAfterComma())
                        printer.space();
                } else
                    printer.newLineAndIndent();
            }
        }
        if (config.isArrayContentsOnSameLine()) {
            if (config.isSpaceWithinBrackets())
                printer.space();
        } else {
            printer.decrementIndentLevel();
            printer.newLineAndIndent();
        }
        printer.print("}");
    }
}

File: src/main/java/net/vpg/vjson/value/JSONArray.kt
/*
 * Copyright 2021 Vaibhav Nargwani
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.vpg.vjson.value

import net.vpg.vjson.pretty.PrettyPrinter
import java.util.stream.Collector

class JSONArray : JSONValue, SerializableArray, JSONContainer<Int>, Iterable<JSONValue> {
    private val list: MutableList<JSONValue>

    constructor() {
        this.list = ArrayList<JSONValue>()
    }

    private constructor(list: List<*>) {
        this.list = list.asSequence().map { it.toJSONValue() }.toMutableList()
    }

    val size: Int
        get() = list.size

    fun isEmpty() = list.isEmpty()

    override fun get(t: Int) = list[t].toJSONValue()

    fun add(index: Int, value: Any?) = also { list.add(index, value.toJSONValue()) }

    fun add(value: Any?) = also { list.add(value.toJSONValue()) }

    fun addAll(values: Collection<*>) = also { values.forEach { add(it) } }

    fun addAll(array: JSONArray) = also { list.addAll(array.list) }

    fun remove(index: Int) = also { list.removeAt(index) }

    fun toList() = list

    fun <T> toList(converter: (JSONValue) -> T) = list.asSequence().map(converter).toMutableList()

    fun stream() = list.stream()

    override fun iterator() = list.iterator()

    override fun deserialize() = list.asSequence().joinToString(",", "[", "]") { it.deserialize() }

    override val type = Type.ARRAY
    override val raw: Any?
        get() = list.asSequence().map { it.raw }.toList()

    override fun toArray() = this

    override fun toPrettyString(printer: PrettyPrinter) {
        val config = printer.config
        printer.print("[")
        printer.nextLine(config.isArrayContentsOnNewLine, +1, config.isSpaceWithinBrackets)
        val iterator = list.iterator()
        while (iterator.hasNext()) {
            iterator.next().toPrettyString(printer)
            if (!iterator.hasNext()) break
            printer.spaceIf(config.isSpaceBeforeComma)
            printer.print(",")
            printer.nextLine(config.isArrayContentsOnNewLine, 0, config.isSpaceAfterComma)
        }
        printer.nextLine(config.isArrayContentsOnNewLine, -1, config.isSpaceWithinBrackets)
        printer.print("]")
    }

    companion object {
        fun List<*>.toJSON() = JSONArray(this)

        fun <T> collector() = Collector.of<T, JSONArray>(
            { JSONArray() },
            { obj, value -> obj.add(value) },
            { obj, array -> obj.addAll(array) }
        )
    }
}
